### PR TITLE
Topic recognition VQ-VAE + PixelCNN for Image Generation on Hip MRI

### PR DESCRIPTION
**Pull request by rapidHarshit**
Hello,
This pull request adds an image generation project using Vector Quantised-Variational Autoencoder (VQ-VAE) with PixelCNN. VQ-VAE learns compressed discrete latent codes, addressing VAE limitations. The PixelCNN model learns over these quantized vector space for generating them from random noise.
By the use of PixelCNN learned probability distribution, the model generates the quantized vector space followed by VQ-VAE's decoder to generate a new Image. 

Key modules include:

`modules.py` for model architecture
`train_VQVAE.py`and `train_PixelCNN.py` for model training.
`predict.py` for generating and comparing images.
`dataset.py` for data loading and filtering inconsistent dimensions.
And other utilities modules to support these main modules.
The project uses 12,460 MRI images and achieves an average SSIM of 0.90 for VQVAE Image reconstruction and more than 0.7 SSIM score for PixelCNN and VQVAE combination. 
Dependencies can managed via the provided environment.yml or requirement.txt.

Workflow:

VQ-VAE Training:
Encoder maps input images to continuous latent vectors, which are quantized into discrete codes.
Decoder reconstructs images from these codes, optimizing reconstruction, codebook, and commitment losses.

PixelCNN Training:
PixelCNN models the prior distribution over the discrete latent codes from VQ-VAE.
It sequentially predicts latent codes, learning dependencies among them.

Image Generation:
PixelCNN samples quantized latent codes based on the learned distribution.
VQ-VAE's decoder generates new images from the sampled codes.

Thanks,
Harshit Rathore
s4814374

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: VQVAE on HipMRI ---- Hard Difficulty

1. Algorithm solves the problem:
Reasonably clear reconstructed images, claimed SSIM over 0.9 without any SSIM output, no data splitting -1.5 (3.5 Marks)

2. Algorithm demonstrated and functions: 
Good latent and quantized space representation (3 Marks)

3. Design:
Modularized design and good programming practice using utils scripts (1 Mark)

4. Commenting:
Meaningful and detailed comments (1 Mark)

5. Algorithm above Normal Difficulty: (5 Marks)

6. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:

- Description on the training process should include key hyperparameters and tuning (e.g. why you ended up choosing LR of 1e-2 for VQVAE, what else you had attempted to improve the performance and what were their implications); You might also want to justify the loss curve(s), I'd recommend having one loss curve throughout all epochs, but this is fine;
- It seems no data splitting, i.e. training, validation, test sets (those sets are available on Rangpur) as you only used `dataloader = get_dataloader("HipMRI_study_keras_slices_data", batch_size = BATCH_SIZE)` for both training and predicting;
- SSIM calculated on the whole dataset in `predict.py`, you could have a plot of SSIM during training along with the loss, merely giving a number of 0.90 in the README.md is not well-supported;
- Images not shown properly on README.md due to incorrect path configuration;
- It's also a good idea to have references in the code.

**Comment by rapidHarshit** [comment]
Hello @Uto0op,

Thank you for the evolution.

Is this is "claimed SSIM over 0.9 without any SSIM output, no data splitting, no validation" what I need to update in my code? Do I need to update the code by following your suggestions?

To resubmit the code and documentation, I need to create a new pull request and also re upload the document?

And what about the rest of the marks regarding the Pull Request (5 Marks), Documentation (10 Marks) and Commit Log (5 Marks)? Will they get evolved after my updates? Or these 20 marks will get converted to the final 30 marks?

Thank you very much for you support.

Regards,
Harshit

**Comment by YashMitttal** [comment]
Some additional feedback:

- The dataset contains images of different dimensions (e.g., 256x128 and 256x144). It would be beneficial to mention how these inconsistencies are handled, beyond simply filtering out the 60 images. Are the images resized or padded? This should be clarified
- Did the training reach convergence at the epoch numbers shown? If so, how many epochs were needed for stable training?
- There is a brief mention of training PixelCNN, but more detail could be provided. 
- Including a section discussing potential improvements or future work would be useful.

To address your question about the marks for pull requests, documentation, and commit logs, we will provide you with the marks separately in some time.

**Comment by YashMitttal** [comment]
Additionally to answer your doubts around ReadMe:

- Yes, you should mention the key hyperparameters used during training (e.g., learning rate, batch size, number of epochs, optimizer type, codebook size, etc.). However, it is not necessary to describe each hyperparameter in exhaustive detail unless you want to clarify how specific values were chosen or their effect on performance.
- It's helpful to briefly explain what the key hyperparameters do.
- if you're looking for SSIM scores during training or validation, it's generally more informative to plot them during training or validation as part of the loss curves. But, do write some information about the result that shows your understanding.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>Data leakage found. No data splits.</td><td>-1</td></tr>
	<tr><td></td><td>Difficulty : Hard. VQVAE (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. Needs more work.</td><td>-2</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Acceptable. No references.</td><td>-1</td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF submitted.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-4</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_
